### ISSUE TITLE
access denied during file renaming on Azure Files share

### ISSUE THREAD

**Issue by ekkelenkamp** [Unidata/tds#347]
We run a thredds server to service netcdf files. In our process we write a netcdf file to a Azure Files share using a tmp prefix. Once the file has been written, we rename it to the actual file name.
We noted that thredds is somehow locking the file resulting in a access denied error. If we stop thredds, no access denied errors occur.
I couldn't find any documentation about file locking done by thredds. Is there any documentation or configuration available to configure the file locking strategy?

**Comment by haileyajohnson** [comment]
What version of the TDS are you running?

**Comment by ekkelenkamp** [comment]
Sorry, should have reported that right away. Version 5.4.

The manifest reports the following:

Manifest-Version: 1.0
Implementation-Title: THREDDS Data Server (TDS)
Implementation-Version: 5.4-SNAPSHOT
Implementation-Vendor-Id: edu.ucar
Implementation-Vendor: UCAR/Unidata
Implementation-URL: https://www.unidata.ucar.edu/software/tds/
Created-By: Gradle 6.9.1
Build-Jdk: 11.0.12
Built-By: ubuntu
Built-On: 2022-04-01T04:26:44+0000

**Comment by haileyajohnson** [comment]
Great, thanks! Could you also verify what process is in fact locking the files? e.g. try `lsof` or similar to see what has it open and let us know?

**Comment by ekkelenkamp** [comment]
It is the tomcat java process. The Only deployed application is thredds.

{code}
COMMAND PID USER   FD   TYPE DEVICE SIZE/OFF            NODE NAME
java      1 fews  220r   REG   0,46  1870195 844424930364399 yaz_drn_grid.nc
[fews@8bd29a1c5e9a modelresults]$ lsof yaz_drn_grid.nc
{code}

It is easily reproducable. I have just one netcdf file in my data folder.
Using the Opendap GUI and selecting a variable and use the GetAscii action, results in a lock. 

![image](https://user-images.githubusercontent.com/7031474/226900487-83e04d8a-b002-45af-a606-bea30b8cd9f6.png)

After I restart tomcat, the lock is removed again.

**Comment by ekkelenkamp** [comment]
Have you been able to reproduce or do you need more information?

**Comment by tdrwenski** [comment]
I can definitely see a file getting a read lock after it has been requested on TDS. This is due to the RandomAccessFile cache. On unix I can rename a file while it has a read lock on it without issue however.

Maybe I did not completely understand your workflow, but do you need TDS to index that file before the rename? Could you instead just not include files with the "tmp" prefix in your catalog?

If that is not an option, you could consider clearing your cache before renaming the files, as that will remove the read lock. This can be done for instance through [the admin page](https://docs.unidata.ucar.edu/tds/5.4/userguide/remote_management_ref.html).

**Comment by ekkelenkamp** [comment]
We have other workflows where this is an issue. 
We merge netcdf files. When one of the files is locked, that fails as well.
We also regenerate existing netcdf files so we need to overwrite them. This fails as well when the file is locked.

On linux the rename seems to succeed, but if you use ls -al you will see it isn't renamed yet. Using the opendap gui, you will see a lot of erros.
So in our case, the lock is really blocking our workflows.

**Comment by tdrwenski** [comment]
Yes, with the caveat that it will affect performance, you can add to your threddsConfig.xml:
```
<RandomAccessFile>
  <maxFiles>0</maxFiles>
</RandomAccessFile>
  ```
  I'm not sure if this is a good permanent fix, however, as I know there are other caches like for WMS that do not currently have a configurable way to to be disabled.

> On linux the rename seems to succeed, but if you use ls -al you will see it isn't renamed yet.

Can you give me an example of the output you see in this situation? I don't think I can reproduce this. What Linux OS are you using?

**Comment by ekkelenkamp** [comment]
Thanks for the suggestion for the RandomAccessFile config. I'll try it out soon.
I also saw the NetcdfFileCache configuration. Might that also be a source of locked files?
Since we don't use the WMS service, is suppose the impact isn't that large.

**Comment by tdrwenski** [comment]
I don't believe the NetcdfFileCache will cause a read lock. Let us know if you still have issues though!

**Comment by craigstanton** [comment]
We're facing the same problem and the slowdown from removing the cache is detrimental to the service we're providing. Aside from restarting TDS we're stuck. Is there an automatable method to flush caches or refresh disk reads? I am hoping there's an API endpoint we could trigger and built into our workflow. In another system we use a request can be made that with the cacheBypass HTTP header or query parameter set to true, it skips the cache on that call, goes to the real file and then stores that result in the cache on the way out. If this was a feature of TDS then as soon as we know the files have been switched out we'd make a request that gets makes TDS read from the new version and all subsequent user requests would too.

**Comment by tdrwenski** [comment]
Would you mind sharing a bit more info about your set up so we could look into it a bit more? For instance, your operating system and if you are using docker or not?

In the mean time, if you are looking for a work around I hope this can be useful for you:

The [admin service](https://docs.unidata.ucar.edu/tds/5.4/userguide/remote_management_ref.html) has endpoints for clearing the caches. You could also try adjusting the cache [settings](https://docs.unidata.ucar.edu/tds/5.4/userguide/tds_config_ref.html), for instance you can say to `scour` the cache every hour down to `minFiles` (you can set `minFiles` to 0 if you want all files scoured).

**Comment by craigstanton** [comment]
We're running it in on Ubuntu 18 in a container from "unidata/thredds-docker:5.4"

We don't have the Tomcat users set up so no admin access yet but I'm going to experiment. 

The only cache related things in threddsConfig.xml are these, with the third one added just now


  <NetcdfSubsetService>
    <allow>true</allow>
    <scour>10 min</scour>
    <maxAge>-1 min</maxAge>
  </NetcdfSubsetService>
  
  <Netcdf4Clibrary>
    <libraryPath>/usr/local/lib</libraryPath>
    <libraryName>netcdf</libraryName>
    <useForReading>true</useForReading>
  </Netcdf4Clibrary>
  
  <RandomAccessFile>
    <maxFiles>0</maxFiles>
  </RandomAccessFile>

**Comment by craigstanton** [comment]
Now the I have enabled the admin pages it looks like this URL could refetched with basic auth and trigger a RandomAccessFile cache clear. I'm not entirely sure if that will release the files from the TDS/Tomcat locks though

http://localhost/thredds/admin/debug?Caches/forceRAFCache

**Comment by tdrwenski** [comment]
Thanks for sending the extra info. I have a few more questions to make sure I understand your situation. Are the files stored on your Ubuntu computer and mounted on the docker container? And does your issue occur when you try to replace a file with a newer file of the same name but cannot due to the lock on the file?

> Now the I have enabled the admin pages it looks like this URL could refetched with basic auth and trigger a RandomAccessFile cache clear. I'm not entirely sure if that will release the files from the TDS/Tomcat locks though
> http://localhost/thredds/admin/debug?Caches/forceRAFCache

I believe the RandomAccessFileCache is the only cache that is causing locks on the files, so clearing that should release the files, though it may have an impact on performance.

**Comment by craigstanton** [comment]
I'm one part of a dispersed team but I think I have all the info now. The files are stored on the same Ubuntu VM that is running Docker and then the folder is mounted into the container as you said. The issue appeared to be when we upload to that server, then rename the original files to a temporary place, rename the new files in to the original location, then delete the temporary place. All done via SFTP.

I'm experimenting now with calling that forceRAFCache page at different stages in the process.

**Comment by tdrwenski** [comment]
@craigstanton were you able to get your set up to work by clearing the random access file cache?

**Comment by tdrwenski** [comment]
We have an improvement in the latest TDS snapshot (https://github.com/Unidata/netcdf-java/pull/1268) that may help some issues related to deleting files that the TDS has locked. Now every time the RandomAccessFile cache is cleaned (set by the scour setting [here](https://docs.unidata.ucar.edu/tds/5.4/userguide/tds_config_ref.html#file-handle-caching), default is 11 min), deleted files will be removed from the cache. If this setting is not enough to fix your specific issue, you can use the workarounds suggested above: change settings for that cache or manually trigger a cleanup of the cache.

I will close this issue but feel free to open a new one if you still have issues.
